Add codec support

jordansissel opened the pull request:
This PR picks up on work from #2 which adds codec support to this output. This PR further makes the following changes on top of that PR:

1) Uses newer `on_event` codec callback that passes the event so we can `sprintf` on the event - required since the `key` setting has historically supported sprintf.
2) Fixes the broken specs. This also included removing some unnecessary specs (that were testing unrelated integration with other plugins) as well as refactoring the specs to use shared_examples.
3) Small refactoring in the output itself.
4) Add necessary dev dependencies to the gemspec so the tests can run.

On my workstation, tests pass: `bundle exec rspec -t redis` under JRuby 1.7.23 (requires redis to be running):

```3 examples, 0 failures```

untergeek commented:
@jordansissel I get an error when I try to run the tests:

```
$ bundle exec rspec -f d -c -t redis
Using Accessor#strict_set for specs
Run options:
  include {:redis=>true}
  exclude {:socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :export_cypher=>true, :integration=>true, :windows=>true}

LogStash::Outputs::Redis
  ship lots of events to a list
    agent(/Users/buh/WORK/logstash-plugins/logstash-output-redis/spec/outputs/redis_spec.rb:30:in) runs (FAILED - 1)
  converts US-ASCII to utf-8 without failures
    agent(/Users/buh/WORK/logstash-plugins/logstash-output-redis/spec/outputs/redis_spec.rb:119:in) runs (FAILED - 2)
  batch mode
    agent(/Users/buh/WORK/logstash-plugins/logstash-output-redis/spec/outputs/redis_spec.rb:74:in) runs (FAILED - 3)

Failures:

  1) LogStash::Outputs::Redis ship lots of events to a list agent(/Users/buh/WORK/logstash-plugins/logstash-output-redis/spec/outputs/redis_spec.rb:30:in) runs
     Failure/Error: pipeline = LogStash::Pipeline.new(config)
     LogStash::PluginLoadingError:
       Couldn't find any input plugin named 'generator'. Are you sure this is correct? Trying to load the generator input plugin resulted in this error: no such file to load -- logstash/inputs/generator
```

It looks like a dependency is missing in the gemspec.

jordansissel commented:
Hmm, are you sure you checked out the right branch? I don't use the `agent` spec block in any specs in this anymore (I removed them from master and replaced with another mechanism)

untergeek commented:
_facepalm_

untergeek commented:
This is working:

```
bundle exec rspec -f d -c -t redis
Using Accessor#strict_set for specs
Run options:
  include {:redis=>true}
  exclude {:socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :export_cypher=>true, :integration=>true, :windows=>true}

LogStash::Outputs::Redis
  when batch_mode is true
    should have batch mode enabled
    should successfully send all events to redis
  when batch_mode is false
    should successfully send all events to redis

Finished in 6.86 seconds (files took 1.32 seconds to load)
3 examples, 0 failures

Randomized with seed 42995
```

Are there only 3 tests?

untergeek commented:
LGTM™

jordansissel commented:
Yeah, some of the tests I removed were more related to integration and regression tests for other bits and bobs (generator input, json parsing, etc) and not specific to this plugin.

The two main tests right now will test both batch and non-batch modes and ensure we send in-order and receive all events that were output. Good enough :)
